### PR TITLE
Improving address generation

### LINKED ISSUE #66: Function register_contract in wasm module should behave more like real blockchain
```
let contract_addr = self.register_contract(
    storage,
    code_id,
    sender.clone(),
    admin.map(Addr::unchecked),
    label,
    block.height,
)?;
```
For the same sender, checksum / code id and salt, the implementation above should generate the same address and fail, just like [here](https://github.com/CosmWasm/wasmd/blob/90d43687ce0facec3cf0c912cf5b331abfdf751c/x/wasm/keeper/keeper.go#L264).

### PR DESCRIPTION
**Pull request by DariuszDepta**
closes #66 

---

## Key changes
- extracted the address generator logic to separate module,
- refactored and improved the interface of `AddressGenerator` trait,
- implemented generation of predictable addresses to have a collision for the same salt,
- added test cases that show how to provide custom `Ap`i and custom `AddressGenerator` that support Bech32 address encoding.

**Users may now provide their own `AddressGenerator` implementation with predictable contract addresses in any address encoding.**

### CODE REVIEW

**Review by chipshort — COMMENTED** [review]
Haven't finished yet, but here are some comments already.
I am warming up more to the idea of doing small breaking changes before 1.0

File: src/addresses.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                None,
+                None,
+                Order::Ascending,
+            )
+            .count();
+        Addr::unchecked(format!("contract{}", count))
+    }
+
+    /// Generates a _non-predictable_ contract address.
+    fn contract_address(
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/addresses.rs]: Yes, good point. I prepared a default implementation just used as a default in `WasmKeeper`. It even simplifies the solution.

File: src/error.rs [hunk 2]
```diff
@@ -27,6 +27,12 @@ pub enum Error {
 
     #[error("code id {0}: no such code")]
     UnregisteredCodeId(u64),
+
+    #[error("Address generator failure: {0}")]
+    AddressGeneratorFailure(String),
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/error.rs]: Right. This is already from "_future_" change, removed for now.

File: src/wasm.rs [hunk 3]
```diff
@@ -824,24 +821,54 @@ where
...
     pub fn register_contract(
         &self,
+        api: &dyn Api,
         storage: &mut dyn Storage,
         code_id: u64,
         creator: Addr,
         admin: impl Into<Option<Addr>>,
         label: String,
         created: u64,
+        salt: impl Into<Option<Binary>>,
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/wasm.rs:838]: Yes, I guess the function `register_contract` (like some other also) should be just private to `WasmKeeper`. This change is braking, but I guess no one should use it in the future.

File: src/addresses.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        &self,
+        api: &dyn Api,
+        storage: &mut dyn Storage,
+        code_id: u64,
+        instance_id: u64,
+    ) -> AnyResult<Addr>;
+
+    /// Generates a _predictable_ contract address.
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/addresses.rs]: By implementing the defaults for this trait, I've improved its documentation.

File: src/error.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chipshort) [reply src/error.rs]: Ah, I see. In that case you could have left it in, but either way is fine.

**Review by chipshort — APPROVED** [review]
LGTM now

**Pull request merged**